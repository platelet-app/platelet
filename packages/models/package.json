{
    "name": "@platelet-app/models",
    "version": "1.0.1",
    "description": "DataStore models for the platelet.app project",
    "homepage": "https://platelet.app",
    "bugs": {
        "url": "https://github.com/platelet-app/platelet/issues"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/platelet-app/platelet.git"
    },
    "license": "Apache-2.0",
    "author": "Theo Cranmore <theo@platelet.app>",
    "type": "module",
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "build": "cd ../.. && amplify api gql-compile && amplify codegen models && cd packages/models && cp ../../models/* dist",
        "prepublishOnly": "npm version patch"
    }
}
