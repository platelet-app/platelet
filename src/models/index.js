// @ts-check
import { initSchema } from '@aws-amplify/datastore';
import { schema } from './schema';

const Role = {
  "USER": "USER",
  "COORDINATOR": "COORDINATOR",
  "RIDER": "RIDER",
  "ADMIN": "ADMIN"
};

const CommentVisibility = {
  "EVERYONE": "EVERYONE",
  "ME": "ME"
};

const TaskStatus = {
  "NEW": "NEW",
  "ACTIVE": "ACTIVE",
  "PICKED_UP": "PICKED_UP",
  "DROPPED_OFF": "DROPPED_OFF",
  "CANCELLED": "CANCELLED",
  "REJECTED": "REJECTED",
  "ABANDONED": "ABANDONED",
  "COMPLETED": "COMPLETED"
};

const Priority = {
  "HIGH": "HIGH",
  "MEDIUM": "MEDIUM",
  "LOW": "LOW"
};

const DeliverableTypeIcon = {
  "BUG": "BUG",
  "CHILD": "CHILD",
  "DOCUMENT": "DOCUMENT",
  "EQUIPMENT": "EQUIPMENT",
  "OTHER": "OTHER"
};

const DeliverableUnit = {
  "NONE": "NONE",
  "LITER": "LITER",
  "MILLILITER": "MILLILITER",
  "GRAM": "GRAM",
  "ITEM": "ITEM",
  "BOX": "BOX"
};

const { User, PossibleRiderResponsibilities, RiderResponsibility, Comment, TaskAssignee, Task, Location, Handover, Deliverable, DeliverableType, VehicleAssignment, Vehicle, Tenant, Statistics, SendFeedback, AddressAndContactDetails, S3Object } = initSchema(schema);

export {
  User,
  PossibleRiderResponsibilities,
  RiderResponsibility,
  Comment,
  TaskAssignee,
  Task,
  Location,
  Handover,
  Deliverable,
  DeliverableType,
  VehicleAssignment,
  Vehicle,
  Tenant,
  Role,
  CommentVisibility,
  TaskStatus,
  Priority,
  DeliverableTypeIcon,
  DeliverableUnit,
  Statistics,
  SendFeedback,
  AddressAndContactDetails,
  S3Object
};