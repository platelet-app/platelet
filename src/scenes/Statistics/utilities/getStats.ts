import { Task, Role, Priority, TaskStatus, User } from "../../../API";

type PriorityUnion = Priority | "NONE" | "TOTAL";

type Stats = {
    common: {
        TOTAL: number;
        [key: string]: number;
    };
    priorities: {
        [key in PriorityUnion]?: number;
    };
    riders: {
        [key: string]: {
            [key: string]: number;
        };
    };
    riderResponsibilities: {
        [key: string]: {
            [key: string]: number;
        };
    };
};

export default function getStats(data: Task[], whoamiId: string) {
    const stats: Stats = {
        common: { TOTAL: 0 },
        priorities: {},
        riders: {},
        riderResponsibilities: {},
    };
    // get all tasks within the date range
    const myTasks = data.filter((t) =>
        t.assignees?.items?.some(
            (a) =>
                !a?._deleted &&
                a?.role === Role.COORDINATOR &&
                a?.assignee?.id === whoamiId
        )
    );
    const myTaskIds = myTasks.map((t) => t.id);
    stats.common.TOTAL = myTasks.length;
    // get the status stats
    for (const status of Object.values(TaskStatus)) {
        if (status === TaskStatus.PENDING) continue;
        stats.common[status] = myTasks.filter(
            (task) => task.status === status
        ).length;
    }
    // get the priority stats
    for (const priority of Object.values(Priority)) {
        stats.priorities[priority] = myTasks.filter(
            (task) => task.priority === priority
        ).length;
    }
    // any unset priorities
    stats.priorities["NONE"] = myTasks.filter((task) => !task.priority).length;
    stats.priorities["TOTAL"] = myTasks.length;

    // get all the rider assignments that intersect with mine
    const riderAssignmentsAll = data.filter((t) =>
        t.assignees?.items?.some((a) => !a?._deleted && a?.role === Role.RIDER)
    );
    const taskAssignments = riderAssignmentsAll.filter((ta) =>
        myTaskIds.includes(ta.id)
    );
    const assignments = taskAssignments
        .map((ta) => ta?.assignees?.items)
        .flat()
        .filter((a) => !a?._deleted && a?.role === Role.RIDER);
    const activeRiders: { [key: string]: User } = {};
    for (const assignment of assignments) {
        if (assignment?.assignee)
            activeRiders[assignment?.assignee?.id] = assignment?.assignee;
    }
    const riders: { [key: string]: any } = {};
    for (const rider of Object.values(activeRiders)) {
        riders[rider.displayName] = {};
        for (const priority of Object.values(Priority)) {
            const count = assignments.filter(
                (assignment) =>
                    assignment?.assignee.id === rider.id &&
                    assignment.task &&
                    assignment.task.priority === priority
            ).length;
            riders[rider.displayName][priority] = count;
        }
        const noPriority = assignments.filter(
            (assignment) =>
                assignment?.assignee.id === rider.id &&
                assignment.task &&
                !assignment.task.priority
        ).length;
        const total = assignments.filter(
            (assignment) => assignment?.assignee.id === rider.id
        ).length;
        riders[rider.displayName]["TOTAL"] = total;
        riders[rider.displayName]["NONE"] = noPriority;
    }
    const responsibilities: { [key: string]: { [key: string]: number } } = {
        None: { Total: 0 },
    };
    const prioritiesTemplate = {
        [Priority.LOW]: 0,
        [Priority.MEDIUM]: 0,
        [Priority.HIGH]: 0,
    };
    for (const task of myTasks) {
        const t = {
            ...task,
        };
        if (!t.riderResponsibility) {
            responsibilities["None"].Total = responsibilities["None"].Total++;
            continue;
        }
        if (!responsibilities[t.riderResponsibility]) {
            responsibilities[t.riderResponsibility] = {
                Total: 1,
                ...prioritiesTemplate,
            };
        } else {
            responsibilities[t.riderResponsibility].Total++;
        }
    }
    for (const resp of Object.keys(responsibilities)) {
        for (const priority of Object.values(Priority)) {
            if (resp === "None") {
                responsibilities[resp][priority] = myTasks.filter(
                    (t) => !t.riderResponsibility && t.priority === priority
                ).length;
                responsibilities[resp]["Total"] +=
                    responsibilities[resp][priority];
            } else {
                if (resp) {
                    responsibilities[resp][priority] = myTasks.filter(
                        (t) =>
                            t.riderResponsibility &&
                            t.riderResponsibility === resp &&
                            t.priority === priority
                    ).length;
                }
            }
        }
    }
    for (const resp of Object.keys(responsibilities)) {
        if (resp === "None") {
            responsibilities[resp]["None"] = myTasks.filter(
                (t) => !t.riderResponsibility && !t.priority
            ).length;
            responsibilities[resp]["Total"] += responsibilities[resp]["None"];
        } else {
            if (resp) {
                responsibilities[resp]["None"] = myTasks.filter(
                    (t) => t.riderResponsibility === resp && !t.priority
                ).length;
            }
        }
    }

    stats.riders = riders;
    stats.riderResponsibilities = responsibilities;

    return stats;
}
