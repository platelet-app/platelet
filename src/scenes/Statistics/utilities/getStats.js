import { DataStore } from "aws-amplify";
import { priorities, tasksStatus, userRoles } from "../../../apiConsts";
import * as models from "../../../models";

export default async function getStats(role, range, whoamiId) {
    const stats = {
        common: {},
        priorities: {},
        riders: {},
        riderResponsibilities: {},
    };
    // get all tasks within the date range
    const tasksWithinRange = await DataStore.query(models.Task, (t) =>
        t.or((t) =>
            t
                .createdAt("eq", undefined)
                .and((t) =>
                    t.createdAt("le", range.start).createdAt("ge", range.end)
                )
        )
    );
    const taskIds = tasksWithinRange.map((t) => t.id);
    const coordAssignmentsAll = await DataStore.query(
        models.TaskAssignee,
        (a) => a.role("eq", userRoles.coordinator)
    );
    // get all of my tasks as a coordinator that fit in the range
    const myCoordAssignments = coordAssignmentsAll.filter(
        (a) => a.task && a.assignee && a.assignee.id === whoamiId
    );
    const myTasks = myCoordAssignments
        .filter((ta) => taskIds.includes(ta.task.id))
        .map((a) => a.task);
    const myTaskIds = myTasks.map((t) => t.id);
    console.log(myTasks);
    console.log(tasksWithinRange);
    stats.common.TOTAL = myTasks.length;
    // get the status stats
    for (const status of Object.values(tasksStatus)) {
        stats.common[status] = myTasks.filter(
            (task) => task.status === status
        ).length;
    }
    // get the priority stats
    for (const priority of Object.values(priorities)) {
        stats.priorities[priority] = myTasks.filter(
            (task) => task.priority === priority
        ).length;
    }
    // any unset priorities
    stats.priorities["NONE"] = myTasks.filter((task) => !task.priority).length;
    stats.priorities["TOTAL"] = myTasks.length;

    // get all the rider assignments that intersect with mine
    const riderAssignmentsAll = await DataStore.query(
        models.TaskAssignee,
        (a) => a.role("eq", userRoles.rider)
    );
    const taskAssignments = riderAssignmentsAll.filter(
        (ta) => ta.task && myTaskIds.includes(ta.task.id)
    );
    const activeRiders = {};
    for (const assignment of taskAssignments) {
        if (assignment.assignee)
            activeRiders[assignment.assignee.id] = assignment.assignee;
    }
    const riders = {};
    for (const rider of Object.values(activeRiders)) {
        riders[rider.displayName] = {};
        for (const priority of Object.values(priorities)) {
            const count = taskAssignments.filter(
                (assignment) =>
                    assignment.assignee &&
                    assignment.assignee.id === rider.id &&
                    assignment.task &&
                    assignment.task.priority === priority
            ).length;
            riders[rider.displayName][priority] = count;
        }
        const noPriority = taskAssignments.filter(
            (assignment) =>
                assignment.assignee &&
                assignment.assignee.id === rider.id &&
                assignment.task &&
                !assignment.task.priority
        ).length;
        const total = taskAssignments.filter(
            (assignment) =>
                assignment.assignee && assignment.assignee.id === rider.id
        ).length;
        riders[rider.displayName]["TOTAL"] = total;
        riders[rider.displayName]["NONE"] = noPriority;
    }
    const responsibilities = { None: { Total: 0 } };
    const prioritiesTemplate = {
        [priorities.low]: 0,
        [priorities.medium]: 0,
        [priorities.high]: 0,
    };
    for (const task of myTasks) {
        const t = {
            ...task,
        };
        if (!t.riderResponsibility) {
            responsibilities["None"].Total = responsibilities["None"].Total++;
            continue;
        }
        if (!responsibilities[t.riderResponsibility]) {
            responsibilities[t.riderResponsibility] = {
                Total: 1,
                ...prioritiesTemplate,
            };
        } else {
            responsibilities[t.riderResponsibility].Total++;
        }
    }
    for (const resp of Object.keys(responsibilities)) {
        for (const priority of Object.values(priorities)) {
            if (resp === "None") {
                responsibilities[resp][priority] = myTasks.filter(
                    (t) => !t.riderResponsibility && t.priority === priority
                ).length;
                responsibilities[resp]["Total"] +=
                    responsibilities[resp][priority];
            } else {
                if (resp) {
                    responsibilities[resp][priority] = myTasks.filter(
                        (t) =>
                            t.riderResponsibility &&
                            t.riderResponsibility === resp &&
                            t.priority === priority
                    ).length;
                }
            }
        }
    }
    for (const resp of Object.keys(responsibilities)) {
        if (resp === "None") {
            responsibilities[resp]["None"] = myTasks.filter(
                (t) => !t.riderResponsibility && !t.priority
            ).length;
            responsibilities[resp]["Total"] += responsibilities[resp]["None"];
        } else {
            if (resp) {
                responsibilities[resp]["None"] = myTasks.filter(
                    (t) => t.riderResponsibility === resp && !t.priority
                ).length;
            }
        }
    }

    stats.riders = riders;
    stats.riderResponsibilities = responsibilities;
    console.log("Generated stats:", stats);

    return stats;
}
