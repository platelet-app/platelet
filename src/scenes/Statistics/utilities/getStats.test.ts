import * as APITypes from "../../../API";
import * as models from "../../../models";
import getStats from "./getStats";

const tenantId = "someTenantId";

describe("getStats", () => {
    test("should return the correct stats", () => {
        const mockWhoami = new models.User({
            tenantId,
            displayName: "someDisplayName",
            roles: [models.Role.ADMIN, models.Role.COORDINATOR],
            username: "someUsername",
            cognitoId: "someCognitoId",
        });
        const mockTask = {
            id: "someId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: "someRiderResponsibility",
            priority: APITypes.Priority.HIGH,
            status: APITypes.TaskStatus.COMPLETED,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        task: {
                            priority: APITypes.Priority.HIGH,
                        },
                        assignee: {
                            id: "someId",
                            displayName: "Rider Name",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: mockWhoami.id,
                            displayName: "Coordinator Name",
                        },
                    },
                ],
            },
            dateCreated: new Date().toISOString().split("T")[0],
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };
        const mockTaskNoPriority = {
            id: "someId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: null,
            priority: null,
            status: APITypes.TaskStatus.NEW,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        task: {
                            priority: null,
                        },
                        assignee: {
                            id: "someId",
                            displayName: "Rider Name",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: mockWhoami.id,
                            displayName: "Coordinator Name",
                        },
                    },
                ],
            },
            dateCreated: new Date().toISOString().split("T")[0],
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };
        const notMyMockTask = {
            id: "anotherId",
            timeOfCall: "2021-05-01T00:00:00.000Z",
            riderResponsibility: "someRiderResponsibility",
            priority: APITypes.Priority.HIGH,
            status: APITypes.TaskStatus.COMPLETED,
            createdAt: new Date().toISOString(),
            assignees: {
                items: [
                    {
                        role: models.Role.RIDER,
                        assignee: {
                            id: "someId",
                            displayName: "Not My Rider Name",
                        },
                    },
                    {
                        role: models.Role.COORDINATOR,
                        assignee: {
                            id: "someOtherId",
                            displayName: "Coordinator Name",
                        },
                    },
                ],
            },
            _version: 1,
            _deleted: null,
            _lastChangedAt: 1620000000000,
        };
        const result = getStats(
            [mockTask, notMyMockTask, mockTaskNoPriority],
            mockWhoami.id
        );
        expect(result).toMatchSnapshot();
    });
});
