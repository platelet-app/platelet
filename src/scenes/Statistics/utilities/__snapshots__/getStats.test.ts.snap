// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStats should return the correct stats 1`] = `
Object {
  "common": Object {
    "ABANDONED": 0,
    "ACTIVE": 0,
    "CANCELLED": 0,
    "COMPLETED": 2,
    "DROPPED_OFF": 0,
    "NEW": 1,
    "PICKED_UP": 0,
    "REJECTED": 0,
    "TOTAL": 3,
  },
  "priorities": Object {
    "HIGH": 2,
    "LOW": 0,
    "MEDIUM": 0,
    "NONE": 1,
    "TOTAL": 3,
  },
  "riderResponsibilities": Object {
    "None": Object {
      "HIGH": 0,
      "LOW": 0,
      "MEDIUM": 0,
      "None": 1,
      "Total": 1,
    },
    "someRiderResponsibility": Object {
      "HIGH": 2,
      "LOW": 0,
      "MEDIUM": 0,
      "None": 0,
      "Total": 2,
    },
  },
  "riders": Object {
    "Rider Name": Object {
      "HIGH": 1,
      "LOW": 0,
      "MEDIUM": 0,
      "NONE": 1,
      "TOTAL": 2,
    },
  },
}
`;
